server/task/src/test/java/org/apache/james/task/MemoryTaskManagerWorkerTest.java

```diff
@@ -0,0 +1,163 @@
+package org.apache.james.task;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+import org.junit.Test;
+
+import reactor.core.publisher.Mono;
+
+public class MemoryTaskManagerWorkerTest {
+
+    private final MemoryTaskManagerWorker worker = new MemoryTaskManagerWorker();
+
+    private final Task successfulTask = () -> Task.Result.COMPLETED;
+    private final Task failedTask = () -> Task.Result.PARTIAL;
+    private final Task throwingTask = () -> {
+        throw new RuntimeException("Throwing Task");
+    };
+
+    @Test
+    public void aSuccessfullTaskShouldCompleteSuccessfully() {
+        assertThatTaskSucceed(successfulTask);
+    }
+
+    @Test
+    public void aFailedTaskShouldCompleteWithFailedStatus() {
+        assertThatTaskFailed(failedTask);
+    }
+
+    @Test
+    public void aThrowingTaskShouldCompleteWithFailedStatus() {
+        assertThatTaskFailed(throwingTask);
+    }
+
+    @Test
+    public void theWorkerShouldReportThatATaskIsInProgress() {
+        TaskId id = TaskId.generateTaskId();
+        CountDownLatch latch = new CountDownLatch(1);
+        CountDownLatch taskLaunched = new CountDownLatch(1);
+
+        Task inProgressTask = () -> {
+            taskLaunched.countDown();
+            await(latch);
+            return Task.Result.COMPLETED;
+        };
+
+        TaskWithId taskWithId = new TaskWithId(id, inProgressTask);
+        TaskExecutionDetails executionDetails = TaskExecutionDetails.from(inProgressTask, id);
+        ConcurrentHashMap<TaskId, TaskExecutionDetails> idToDetails = new ConcurrentHashMap<>();
+        idToDetails.put(id, executionDetails);
+
+        worker.executeTask(taskWithId, updateDetails(idToDetails, id)).cache();
+        await(taskLaunched);
+        assertThat(idToDetails.get(id).getStatus()).isEqualTo(TaskManager.Status.IN_PROGRESS);
+        latch.countDown();
+    }
+
+    @Test
+    public void theWorkerShouldNotRunACancelledTask() {
+        TaskId id = TaskId.generateTaskId();
+        AtomicInteger counter = new AtomicInteger(0);
+
+        Task task = () -> {
+            counter.incrementAndGet();
+            return Task.Result.COMPLETED;
+        };
+
+        TaskWithId taskWithId = new TaskWithId(id, task);
+        TaskExecutionDetails executionDetails = TaskExecutionDetails.from(task, id);
+        ConcurrentHashMap<TaskId, TaskExecutionDetails> idToDetails = new ConcurrentHashMap<>();
+
+        idToDetails.put(id, executionDetails.cancel());
+
+        worker.executeTask(taskWithId, updateDetails(idToDetails, id)).cache();
+
+        assertThat(counter.get()).isEqualTo(0);
+        assertThat(idToDetails.get(id).getStatus()).isEqualTo(TaskManager.Status.CANCELLED);
+    }
+
+    @Test
+    public void theWorkerShouldCancelAnInProgressTask() {
```

Comment:
can you add a test about:
 - canceling a cancelled task
 - canceling a failed task
 - canceling a completed task

Reply:
idem it's the manager responsability. 
The canceling a cancelled task already exists  : cancelShouldBeIdempotent

i will add the two others in the manager unit test

Reply:
Added the missing tests in the manager test.